bridge/bridge/deploy/local.py

```diff
@@ -0,0 +1,283 @@
+                "mlflow",
+                "models",
+                "serve",
+                "--model-uri",
+                self.model_uri,
+                "--port",
+                str(self.port),
+                "--host",
+                "127.0.0.1",
+            ],
+            stdout=stdout,
+            stderr=stderr,
+            preexec_fn=os.setsid,
+            env=env,
+        )
+
+    def destroy(self):
+        if self.serve_handle is not None:
+            os.killpg(os.getpgid(self.serve_handle.pid), signal.SIGTERM)
+            self.serve_handle.wait()
+            self.serve_handle = None
+
+
+class LocalDeployTarget(DeployTarget):
+    target_name = "local"
+
+    @property
+    def target_id(self) -> str:
+        return socket.gethostname()
+
+    def list_models(self) -> List[Model]:
+        # Run reconciliation loop for deployments
+        mvs = list(self.running_models.keys())
+        for mv in mvs:
+            self.running_models[mv].check()
+
+        result = []
+        for model_name, d in self.routes.items():
+            model = Model(name=model_name, versions={})
+            for stage, port in d.items():
+                for mv, md in self.running_models.items():
+                    if port == md.port:
+                        model.versions[stage] = set([mv])
+                        break
+            result.append(model)
+        logger.info(result)
+        return result
+
+    def teardown(self):
+        # TODO: Shutdown Flask
+        self.routes = {}
+        models = list(self.running_models.keys())
+        for model in models:
+            self.running_models[model].destroy()
+            del self.running_models[model]
+
+    def __init__(self):
+        self.running_models: Dict[ModelVersion, ModelDeployment] = {}
+        # model, stage, port
+        self.routes: Dict[str, Dict[str, int]] = {}
+        self.proxy = LocalDeploymentProxy(self)
+        self.proxy.run()
+
+    def init(self):
+        pass
+
+    def create_versions(self, new_versions: Dict[ModelVersion, Artifact]):
+        for model_version, artifact in new_versions.items():
+            if model_version in self.running_models:
+                logger.warning(
+                    f"Model version already deployed: {model_version}"
+                )
+                continue
+            logger.info(f"Deploying version {model_version}")
+            with socket.socket() as s:
+                s.bind(("", 0))
+                port = s.getsockname()[1]
+            logger.info(f"Selected port {port}")
+            model_dir = os.path.splitext(os.path.splitext(artifact.path)[0])[0]
+            self.running_models[model_version] = ModelDeployment(
+                model_version, model_dir, port
+            )
+
+    def update_version_stage(
+        self,
+        current_routing: Dict[str, Dict[str, Set[str]]],
+        desired_routing: Dict[str, Dict[str, Set[str]]],
+    ):
+        self.routes = defaultdict(dict)
```

Comment:
This should be in the init function instead of the assignment of a regular `{}` empty dict

Reply:
Ah I see what you're doing. Because you have conflated the existence of a version and the running of that version, you do the init/teardown in the create/delete version calls. So all you really need to do in this function is rebuild the routing based on the desired state. Is that right?

Reply:
Yes, for this target, if a "model version" exists, it is running. We just need to tell Flask to route traffic to it. This seems simpler to me.

Reply:
Yep I think it is smart